health/health.go

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2024 The VeChainThor developers
+
+// Distributed under the GNU Lesser General Public License v3.0 software license, see the accompanying
+// file LICENSE or <https://www.gnu.org/licenses/lgpl-3.0.html>
+
+package health
+
+import (
+	"sync"
+	"time"
+
+	"github.com/vechain/thor/v2/thor"
+)
+
+type BlockIngestion struct {
+	BestBlock          *thor.Bytes32 `json:"bestBlock"`
+	BestBlockTimestamp *time.Time    `json:"bestBlockTimestamp"`
+}
+
+type Status struct {
+	Healthy        bool            `json:"healthy"`
+	BlockIngestion *BlockIngestion `json:"blockIngestion"`
+	ChainSync      bool            `json:"chainSync"`
+}
+
+type Health struct {
+	lock         sync.RWMutex
+	newBestBlock time.Time
+	bestBlockID  *thor.Bytes32
+	chainSynced  bool
+}
+
+func (h *Health) NewBestBlock(ID thor.Bytes32) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	h.newBestBlock = time.Now()
+	h.bestBlockID = &ID
+}
+
+func (h *Health) Status() (*Status, error) {
+	h.lock.RLock()
+	defer h.lock.RUnlock()
+
+	blockIngest := &BlockIngestion{
+		BestBlock:          h.bestBlockID,
+		BestBlockTimestamp: &h.newBestBlock,
+	}
+
+	// todo review time slots
+	healthy := time.Since(h.newBestBlock) >= 10*time.Second &&
+		h.chainSynced
+
+	return &Status{
+		Healthy:        healthy,
+		BlockIngestion: blockIngest,
+		ChainSync:      h.chainSynced,
+	}, nil
+}
+
+func (h *Health) ChainSynced() {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	h.chainSynced = true
```

Comment:
Good point, I've modified the ChainSync to ChainSyncStatus and marked as unhealthy if it's syncing.

There is a concern that this may make the ChainSync Status to flicker from Sync to UnSync a bit too often.

Reply:
Maybe we could introduce some kind of time delay before switching state; I wouldn't know exactly how much, but that would make the switch less sudden. Would that be feasible?

Reply:
That's a good idea, will implement, want to try this out with testnet/mainnet first to see if there is a flickering.

Reply:
It doesn't seem to flicker but it's not working as intended. I have a few ideas that I would like to run by you, will ping you tomorrow 👍

Reply:
Modified the healthcheck to bootstrapStatus from chainSync .